src/routing/comm.rs

```diff
@@ -104,9 +112,20 @@ impl Comm {
         recipient: &SocketAddr,
         msg: Bytes,
     ) -> Result<(), SendError> {
-        if let Some(conn) = self.endpoint.get_connection(recipient) {
-            if let Err(err) = conn.send_uni(msg).await {
-                error!("Sending message to {} failed: {}", recipient, err);
+        // This will attempt to used a cached connection
+        if self
+            .endpoint
+            .send_message(msg.clone(), recipient)
+            .await
+            .is_err()
+        {
+            // If the sending of a message failed the connection would no longer
+            // exist in the pool. Calling send_message agail will use a new connection.
+            if let Err(err) = self.endpoint.send_message(msg, recipient).await {
+                error!(
+                    "Sending message to client {:?} failed with error {:?}",
+                    recipient, err
+                );
```

Comment:
This is not quite right. The idea was that we would send the message on a connection that already exists or fail otherwise. But with this change we are establishing new connection on failure which is not going to work for clients in the general case.

Reply:
Although... the reason this didn't work before was that clients weren't listening for incoming connections. But after upgrading clients to use the new qp2p as well, that's not longer the case, is it? So maybe this works in the end.

Reply:
But then this makes this function almost identical to the `send_to` below, except here we also log the error. Could we at least remove the duplication somehow?

Reply:
Ah I see, this is for clients. I can change this :+1: